src/brod_group_controller.erl

```diff
@@ -0,0 +1,889 @@
+      , oCReqV2Topic_L = Offsets
+      },
+  Rsp = send_sync(SockPid, Req),
+  #kpro_OffsetCommitResponse{oCRspTopic_L = Topics} = Rsp,
+  lists:foreach(
+    fun(#kpro_OCRspTopic{topicName = Topic, oCRspPartition_L = Partitions}) ->
+      lists:foreach(
+        fun(#kpro_OCRspPartition{partition = Partition, errorCode = EC}) ->
+          kpro_ErrorCode:is_error(EC) andalso
+            begin
+              log(State, error,
+                  "failed to commit offset for topic=~s, partition=~p\n"
+                  "~p:~s", [Topic, Partition, EC, kpro_ErrorCode:desc(EC)]),
+              erlang:error(EC)
+            end
+        end, Partitions)
+    end, Topics),
+  {ok, State#state{acked_offsets = []}}.
+
+-spec assign_partitions(#state{}) ->
+        [kpro_GroupAssignment()] | no_return().
+assign_partitions(State) when ?IS_LEADER(State) ->
+  #state{ client                        = Client
+        , topics                        = Topics
+        , members                       = Members
+        , partition_assignment_strategy = Strategy
+        } = State,
+  AllPartitions =
+    [ {Topic, Partition}
+      || Topic <- lists:usort(Topics),
+         Partition <- get_partitions(Client, Topic)
+    ],
+  Assignments = do_assign_partitions(Strategy, Members, AllPartitions),
+  lists:map(
+    fun({MemberId, Topics_}) ->
+      PartitionAssignments =
+        lists:map(fun({Topic, Partitions}) ->
+                    #kpro_ConsumerGroupPartitionAssignment
+                      { topicName   = Topic
+                      , partition_L = Partitions
+                      }
+                  end, Topics_),
+      #kpro_GroupAssignment
+        { memberId = MemberId
+        , memberAssignment =
+            #kpro_ConsumerGroupMemberAssignment
+              { version = ?BROD_CONSUMER_GROUP_PROTOCOL_VERSION
+              , consumerGroupPartitionAssignment_L = PartitionAssignments
+              }
+        }
+    end, Assignments);
+assign_partitions(#state{}) ->
+  %% only leader can assign partitions to members
+  [].
+
+-spec get_partitions(client(), topic()) -> [partition()] | no_return().
+get_partitions(Client, Topic) ->
+  Count = ?ESCALATE(brod_client:get_partitions_count(Client, Topic)),
+  lists:seq(0, Count-1).
+
+-spec do_assign_partitions(partition_assignment_strategy(),
+                           [kpro_GroupMemberMetadata()],
+                           [{topic(), partition()}]) -> [member_assignment()].
+do_assign_partitions(roundrobin, Members, AllPartitions) ->
+  %% round robin, we only care about the member id
+  F = fun(#kpro_GroupMemberMetadata{memberId = MemberId}) -> MemberId end,
+  MemberIds = lists:map(F, Members),
+  roundrobin_assign_loop(AllPartitions, MemberIds).
+
+-spec roundrobin_assign_loop([{topic(), partition()}], [member_id()]) ->
+        [member_assignment()].
+roundrobin_assign_loop([], Members) ->
+  %% remove the ones that has no assignments
+  lists:filter(fun(M) -> is_tuple(M) end, Members);
```

Comment:
see spec of assign_partition.
the input list is a list of member_id() which is binary().
the output is a list of {member_id(), [...]}, but the ones that have no assignment will be kept the untouched, i.e. still a member_id().